### PR TITLE
refactor(engine/sink): introduce SinkOutput abstraction

### PR DESCRIPTION
**Pull request by bnimit**
# refactor(engine/sink): introduce SinkOutput abstraction

> **Branch:** `chore/engine-sink-output-abstraction` (22 commits ahead of `main`)
> **Companion future work:** PR2 will add the sandbox security check inside `SinkOutput::from_path` (designed but not started).

## Why

Every sink writer in `engine/runtime/action-sink/src/file/` duplicated the same boilerplate for URI parsing, storage backend acquisition, and bytes write inside its `finish()` (and sometimes `process()`) method:

1. `Uri::from_str(path)` → URI
2. `storage_resolver.resolve(&uri)` → storage backend
3. **Format-specific serialization** (the only step that varies)
4. `storage.put_sync(path, bytes)`

14 sinks × 3-of-4 identical steps = duplicated code with no single chokepoint. This PR extracts steps 1, 2, and 4 into one type, `SinkOutput`, and migrates every sink to it. Each sink now only owns its serialization logic and its own expression evaluation.

## What changes

**New type:** `engine/runtime/action-sink/src/output.rs` — `SinkOutput`. **Storage-focused only**, no knowledge of expression evaluation:

- `from_path(&NodeContext, &str) -> Result<Self, BoxedError>` — parses URI, eagerly acquires storage backend.
- `uri(&self) -> &Uri` — read the resolved URI.
- `write(&self, Bytes) -> Result<(), BoxedError>` — write via the backend.
- `join(&self, &str) -> Result<Self, BoxedError>` — derive child URIs (for groups, tiles, sidecars). Reuses the parent's storage `Arc`.

`Clone + Debug` are derived so sinks can store instances in their buffers.

**Expression evaluation stays in each sink.** The `scope.eval::<String>(...).unwrap_or_else(...)` dance lives at each call site — same place it did before the refactor began. This is a deliberate choice (see "Updates from review" below).

**14 sink migrations** — one commit per sink, simplest → most complex (Zip, Geopackage, CityGml, Json, Xml, Csv, Excel, GeoJson, Czml, Obj, Shapefile, Gltf, Mvt, Cesium3DTiles). Each commit is independently CI-green.

## What does NOT change

- **No new workflow features.** Same `output:` param semantics for every sink.
- **No fixture / YAML edits.** All 83 workflow fixtures referencing `workerArtifactPath` are untouched (`git diff main -- engine/runtime/examples/fixture engine/runtime/tests/fixture engine/testing/data` is empty).
- **No sandbox / security check.** That's the explicit PR2 scope.
- **No Cesium3DTiles `output`/`compressedOutput` consolidation.** Cesium keeps two independent `SinkOutput` instances. PR2 collapses them.
- **No cross-crate signature changes** in `action-processor`. Where a `pub fn` is shared between sink and processor crates, only the sink-local caller migrated (see CityGML below).
- **Expression evaluation semantics unchanged** — same Rhai eval with silent-string fallback in each sink, just no longer abstracted into `SinkOutput`.

## Tests

| Suite | Before | After |
| --- | --- | --- |
| `cargo test -p reearth-flow-action-sink --lib` | 48 | 56 (48 prior + 8 new `SinkOutput` tests: 3 happy paths + 5 edge cases) |
| `cargo make test-rs` (full workspace) | passes | passes |
| Fixture YAML diff vs main | — | empty |

Every existing sink test passes **unmodified**. That's the load-bearing claim of a pure-refactor PR.

The 5 edge-case tests cover: invalid URI rejection with error context, absolute-subpath rejection in `join`, current `..` traversal behavior (pinned for PR2 to verify the sandbox change), `Clone` sharing the storage `Arc`, and a compile-time `Send + Sync + Clone + Debug` assertion that guards future field additions to `SinkOutput`.

**Notable finding from the edge-case tests:** `base.join("../escape.txt")` **currently succeeds and resolves outside the base directory**. This is the exact traversal vulnerability PR2's sandbox check exists to close. The pinned test in this PR makes PR2's behavior change visible.

---

## Updates from review

- **Narrowed `SinkOutput` to URI + storage only.** An earlier iteration had `from_expr` and `evaluate_uri` methods that owned the eval-fallback dance. Per reviewer feedback, those were removed so the type stays decoupled from the expression system — different sinks can evolve their expression handling independently (e.g., when the silent-string fallback is eventually deprecated in favor of an explicit literal-path type, individual sinks can migrate without touching the shared abstraction).
- **Hot-path optimization preserved.** Dynamic-per-feature sinks (JSON/XML/CSV/Excel) use `Entry::Occupied`/`Vacant` so `SinkOutput::from_path` (with its `storage_resolver.resolve` call) only fires when a new output URI appears in the buffer, not on every feature.
- **GLTF regression caught and fixed.** An early iteration moved GltfWriter's output evaluation from `build()` to `finish()`, which silently broke two things: (1) `with:` workflow parameters were no longer injected into the eval scope, so any expression referencing a `with` variable would fail to resolve, and (2) eval failures changed from a build-time `SinkError::BuildFactory` (workflow startup abort) to a silent fallback to the raw expression string. Both reverted — GltfWriter is back to its original semantics, with `SinkOutput::from_path` called in `finish()` from the already-resolved `Uri`.

## Editorial improvements worth calling out

While migrating, a few sinks got small improvements beyond a byte-exact swap. Worth reading so reviewers don't miss them.

### 1. CityGML — partial migration

`write_citygml_to_storage` is a `pub fn` shared between the sink crate and the `action-processor` crate. Migrating its signature would touch a second crate, out of scope here. Only the sink-internal callsite was migrated; the function still receives `&Uri` + `&Arc<StorageResolver>` as before. **Texture sidecar writes still go through the old direct-storage path** and will need PR2 to migrate them when the sandbox check lands.

### 2. Shapefile — `storage.delete_sync` → `std::fs::remove_file`

In the null-shape overwrite cleanup, two `storage.delete_sync(...)` calls were replaced with `std::fs::remove_file(...)`. Rationale: the shapefile sink is **local-filesystem-only by design** — `shapefile::Writer::from_path` is hard-coded to local paths, so the rest of the sink can never see a non-`file://` URI. The two paths are equivalent in practice.

### 3. MVT and Cesium3DTiles — silent `if let Ok` on compress resolve → logged `match`

The compress-output write block originally swallowed storage-resolver failures silently via `if let Ok(storage) = ctx.storage_resolver.resolve(...)`. The refactor changed this to an explicit `match` that emits a `tracing` warn via `event_hub` on the `Err` arm. Upstream behavior preserved — errors are still non-fatal, the sink continues. Operators will see a log line in monitoring when compress resolve fails. Net: latent bug surfacing, no functional regression.

---

## Why this PR is the foundation for PR2

PR2's sandbox security check (rejecting writes outside a per-job allowed root) needs **one place** to attach. Before this PR, that one place didn't exist — every sink had its own copy of the URI → storage acquisition logic, and a sandbox check would need to be duplicated 14 times.

After this PR, the sandbox check is a single conditional inside `SinkOutput::from_path`. Every sink that routes through `SinkOutput` inherits it the moment it lands. That's why the refactor is worth its size.

---

## Reviewer checklist

- [ ] Read `output.rs` first — it's the foundational API everything else routes through. ~80 lines.
- [ ] Spot-check one of the simpler sink migrations (`zip.rs` is good) to see the typical shape.
- [ ] Spot-check one of the dynamic per-feature ones (`json.rs`) to see the `Entry::Occupied`/`Vacant` buffer pattern.
- [ ] Read the "Editorial improvements" notes above and confirm you're comfortable with the small deviations.
- [ ] Verify `git diff main -- engine/runtime/examples/fixture engine/runtime/tests/fixture engine/testing/data` is empty (no fixture changes).
- [ ] Optional: run `cargo make test-qc` and `cargo make test-conv` locally — these exercise sinks end-to-end through real workflow fixtures.

### CODE REVIEW

File: engine/runtime/action-sink/src/output.rs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    /// Construct a `SinkOutput` from a fully-formed path string.
+    ///
+    /// The string must parse as a valid URI (`file://...`, `gs://...`, etc.).
+    /// Resolves the storage backend eagerly. Used by sinks that pre-compile
+    /// Rhai expression ASTs and evaluate them at write time.
+    pub fn from_path(ctx: &NodeContext, path: &str) -> Result<Self, BoxedError> {
+        let resolved = Uri::from_str(path).map_err(|e| -> BoxedError {
+            format!("SinkOutput: invalid path {:?}: {e}", path).into()
+        })?;
+        let storage = ctx.storage_resolver.resolve(&resolved)?;
```
[the remark this answers is not in the document]
Comment (bnimit) [reply engine/runtime/action-sink/src/output.rs]: Good catch — fixed in 29952bd6. The `storage_resolver.resolve` error now wraps the resolved URI in the message: `SinkOutput: failed to resolve storage for {uri}: {e}`.

File: engine/runtime/action-sink/src/output.rs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    resolved: Uri,
+    storage: Arc<Storage>,
+}
+
+impl SinkOutput {
+    /// Construct a `SinkOutput` from a fully-formed path string.
+    ///
+    /// The string must parse as a valid URI (`file://...`, `gs://...`, etc.).
+    /// Resolves the storage backend eagerly. Used by sinks that pre-compile
+    /// Rhai expression ASTs and evaluate them at write time.
```
[the remark this answers is not in the document]
Comment (bnimit) [reply engine/runtime/action-sink/src/output.rs]: Fixed in 29952bd6. Updated the docstring to say it accepts either a URI or a plain filesystem path (treated as `file://` by `Uri::from_str`).

File: engine/runtime/action-sink/src/output.rs [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+                    sub.uri().path().as_path(),
+                    expected.as_path(),
+                    "traversal currently resolves to parent — PR2 will block this"
+                );
+            }
+            Err(e) => {
+                // Today: Uri::join rejects `..` segments outright.
+                let _ = e; // pin that traversal is rejected at the Uri layer
+            }
+        }
```
[the remark this answers is not in the document]
Comment (bnimit) [reply engine/runtime/action-sink/src/output.rs]: Agreed — fixed in 29952bd6. The test now asserts the `Ok` branch deterministically with the expected resolved path, and the "run locally" comment is gone. PR2 will flip this from `expect(...)` to a rejection assertion when the sandbox lands.

File: engine/runtime/action-sink/src/file/json.rs [hunk 4]
```diff
@@ -99,38 +98,30 @@ impl Sink for JsonWriter {
...
-        let uri = Uri::from_str(&path)?;
-        self.buffer.entry(uri).or_default().push(ctx.feature);
+        let node_ctx: NodeContext = ctx.clone().into();
+        let out = SinkOutput::from_expr(&node_ctx, &self.params.output)
+            .map_err(|e| SinkError::JsonWriter(e.to_string()))?;
+        self.buffer
+            .entry(out.uri().clone())
+            .or_insert_with(|| (out, Vec::new()))
+            .1
+            .push(ctx.feature);
```
[the remark this answers is not in the document]
Comment (bnimit) [reply engine/runtime/action-sink/src/file/json.rs]: Good point — fixed in 3a493697. Added `SinkOutput::evaluate_uri` which evaluates the `Expr` to `(path, Uri)` without acquiring storage. `process()` now does the cheap URI parse first, then uses `Entry::Occupied`/`Vacant` to only call `SinkOutput::from_path` (the resolve cost) when the buffer entry is genuinely new.

File: engine/runtime/action-sink/src/file/xml.rs [hunk 5]
```diff
@@ -96,30 +94,28 @@ impl Sink for XmlWriter {
...
-        let uri = Uri::from_str(&path)?;
-        self.buffer.entry(uri).or_default().push(ctx.feature);
+        let node_ctx: NodeContext = ctx.clone().into();
+        let out = SinkOutput::from_expr(&node_ctx, &self.params.output)
+            .map_err(|e| SinkError::XmlWriter(e.to_string()))?;
+        self.buffer
+            .entry(out.uri().clone())
+            .or_insert_with(|| (out, Vec::new()))
+            .1
+            .push(ctx.feature);
```
[the remark this answers is not in the document]
Comment (bnimit) [reply engine/runtime/action-sink/src/file/xml.rs]: Same fix as the JSON case — addressed in 3a493697 via the new `SinkOutput::evaluate_uri` helper and an `Entry::Occupied`/`Vacant` match in `process()`. Storage resolution now only happens on first insert for each output URI.

File: engine/runtime/action-sink/src/file/csv.rs [hunk 6]
```diff
@@ -121,26 +119,24 @@ impl Sink for CsvWriter {
...
-        let uri = Uri::from_str(&path)?;
-        self.buffer.entry(uri).or_default().push(ctx.feature);
+        let node_ctx: NodeContext = ctx.clone().into();
+        let out = SinkOutput::from_expr(&node_ctx, &self.params.output)
+            .map_err(|e| SinkError::CsvWriter(e.to_string()))?;
+        self.buffer
+            .entry(out.uri().clone())
+            .or_insert_with(|| (out, Vec::new()))
+            .1
+            .push(ctx.feature);
```
[the remark this answers is not in the document]
Comment (bnimit) [reply engine/runtime/action-sink/src/file/csv.rs]: Fixed in 3a493697 using the same pattern as the JSON/XML/Excel sinks: `SinkOutput::evaluate_uri` for the cheap path-and-URI parse, then `Entry::Occupied`/`Vacant` to only construct the full `SinkOutput` when inserting.

File: engine/runtime/action-sink/src/file/excel_writer.rs [hunk 7]
```diff
@@ -98,24 +96,23 @@ impl Sink for ExcelWriter {
...
-        let uri = Uri::from_str(&path)?;
-        self.buffer.entry(uri).or_default().push(ctx.feature);
+        let node_ctx: NodeContext = ctx.clone().into();
+        let out = crate::SinkOutput::from_expr(&node_ctx, &self.params.output)
+            .map_err(|e| SinkError::ExcelWriterFactory(e.to_string()))?;
+        self.buffer
+            .entry(out.uri().clone())
+            .or_insert_with(|| (out, Vec::new()))
+            .1
+            .push(ctx.feature);
```
[the remark this answers is not in the document]
Comment (bnimit) [reply engine/runtime/action-sink/src/file/excel_writer.rs]: Fixed in 3a493697. `process()` now uses `SinkOutput::evaluate_uri` for the cheap `(path, Uri)` parse and only calls `SinkOutput::from_path` (the storage-resolve step) when the `Uri` is first inserted into the buffer.

**Review by n4to4 — APPROVED** [review]
LGTM 👍

File: engine/runtime/action-sink/src/file/cesium3dtiles/sink.rs [hunk 8]
```diff
@@ -217,13 +216,18 @@ impl Cesium3DTilesWriter {
         let path = scope
             .eval_ast::<String>(&output)
             .map_err(|e| SinkError::Cesium3DTilesWriter(format!("{e:?}")))?;
-        let output = Uri::from_str(path.as_str()).map_err(SinkError::cesium3dtiles_writer)?;
+        let node_ctx = NodeContext::from(ctx.clone());
+        let output_sink = crate::SinkOutput::from_path(&node_ctx, path.as_str())
```
Comment (asrcpq) [inline engine/runtime/action-sink/src/file/cesium3dtiles/sink.rs]: Since evaluation is isolated from SinkOutput, it is probably not necessary to construct SinkOutput every feature, leading to storage resolution called every feature which is a potential performance issue. MVT writer implementation at L224 seems correct (directly construct URI)

**Review by asrcpq — APPROVED** [review]

File: engine/runtime/action-sink/src/file/cesium3dtiles/sink.rs (hunk 8 above)
Comment (bnimit) [reply engine/runtime/action-sink/src/file/cesium3dtiles/sink.rs]: Good catch! Let me fix that before I resolve the commit signing so that the fix is merged along too!
